test/unit/integrated_function_test.C

```diff
@@ -0,0 +1,309 @@
+              GRINS::IntegratedFunction<libMesh::FunctionBase<libMesh::Real> > integ_func((unsigned int)3,new libMesh::ParsedFunction<libMesh::Real>(functions[i]),rayfire,"integrated_function");
+              comp_qoi.add_qoi(integ_func);
+            }
+
+          comp_qoi.init(*_input,*system);
+          system->attach_qoi(&comp_qoi);
+
+          libMesh::MeshRefinement mr(system->get_mesh());
+
+          unsigned int num_converged = 0;
+          unsigned int iter = 0;
+
+          // limit iterations to prevent excessive runtime
+          unsigned int max_iter = 7;
+          do
+            {
+              std::vector<libMesh::dof_id_type> elems_in_rayfire;
+              rayfire->elem_ids_in_rayfire(elems_in_rayfire);
+
+              system->assemble_qoi();
+
+              libMesh::Real h = -1.0;
+              for (unsigned int i=0; i<elems_in_rayfire.size(); i++)
+                {
+                  libMesh::Real l = (rayfire->map_to_rayfire_elem(elems_in_rayfire[i]))->length(0,1);
+                  if (l>h)
+                      h=l;
+                }
+
+              CPPUNIT_ASSERT(h > 0.0);
+
+              for (unsigned int i=0; i<calc_answers.size(); i++)
+                {
+                  if (!converged[i])
+                    {
+                      libMesh::Real err = std::abs( _sim->get_qoi_value(i) - calc_answers[i] );
+                      if (err < 2.5e-9)
+                        {
+                          num_converged++;
+                          converged[i] = true;
+                          h_vals[i].push_back(std::log10(h));
+                          errors[i].push_back(std::log10(err));
+                        }
+                      else
+                        {
+                          h_vals[i].push_back(std::log10(h));
+                          errors[i].push_back(std::log10(err));
+                        }
+                    }
+                }
+
+              if (num_converged < calc_answers.size())
+                {
+                  if (iter++ > max_iter+1 )
+                    libmesh_error_msg("Exceeded maximum iterations");
+
+                  for (unsigned int i=0; i<elems_in_rayfire.size(); i++)
+                    system->get_mesh().elem(elems_in_rayfire[i])->set_refinement_flag(libMesh::Elem::RefinementState::REFINE);
+
+                  mr.refine_elements();
+                  for (unsigned int i=0; i<elems_in_rayfire.size(); i++)
+                    CPPUNIT_ASSERT( !( system->get_mesh().elem(elems_in_rayfire[i])->active() ) );
+
+                  rayfire->reinit(system->get_mesh());
+                  es->reinit();
+                }
+
+            } while(num_converged < calc_answers.size());
+
+            for (unsigned int i=0; i<functions.size(); i++)
+              {
+                libMesh::Real convergence_rate = this->linear_slope(h_vals[i],errors[i],0,errors[i].size()-1);
+                CPPUNIT_ASSERT_DOUBLES_EQUAL(4.0,convergence_rate,0.08);
```

Comment:
Can we abstract this out some how into the `RegressionHelper` to have a test for the rate?